File: frontend/src/utils/chain.ts
import type { DomainRecord } from '../types/domain';

// ChainLookup indexes the release chain (batch -> sample -> method -> review)
// by business code so pages can render relations and blocking reasons from
// fresh backend data after every reload.
export interface ChainLookup {
  batches: Record<string, DomainRecord>;
  samples: Record<string, DomainRecord>;
  methods: Record<string, DomainRecord>;
}

export const EMPTY_CHAIN: ChainLookup = { batches: {}, samples: {}, methods: {} };

export function buildChainLookup(batches: DomainRecord[], samples: DomainRecord[], methods: DomainRecord[]): ChainLookup {
  const index = (records: DomainRecord[]) => Object.fromEntries(records.map((record) => [record.code, record]));
  return { batches: index(batches), samples: index(samples), methods: index(methods) };
}

export function methodUsable(status: string | undefined): boolean {
  return status === 'active';
}

export function openSamplesOf(batchCode: string, samples: DomainRecord[]): DomainRecord[] {
  return samples.filter((sample) => sample.batchCode === batchCode && sample.status !== 'disposed');
}

// transitionBlockers mirrors the backend release gates so the workbench can
// explain why an action is blocked; the backend remains the authority.
export function transitionBlockers(configKey: string, item: DomainRecord, target: string | null, chain: ChainLookup, actor: string): string[] {
  if (!target) return [];
  const blockers: string[] = [];
  if (configKey === 'labSample' && target === 'accepted') {
    const batch = item.batchCode ? chain.batches[item.batchCode] : undefined;
    if (!batch) blockers.push(`所属批次 ${item.batchCode || '未关联'} 不存在`);
    else if (batch.status !== 'received') blockers.push(`批次 ${batch.code} 状态为 ${batch.status}，须先收货`);
    const method = item.methodCode ? chain.methods[item.methodCode] : undefined;
    if (!method) blockers.push(`检测方法 ${item.methodCode || '未关联'} 不存在`);
    else if (!methodUsable(method.status)) blockers.push(`方法 ${method.code} 版本状态为 ${method.status}，须为 active`);
  }
  if (configKey === 'samplingBatch' && target === 'closed') {
    const open = openSamplesOf(item.code, Object.values(chain.samples));
    if (open.length) blockers.push(`仍有 ${open.length} 个样本未处置：${open.slice(0, 3).map((sample) => sample.code).join('、')}${open.length > 3 ? ' 等' : ''}`);
  }
  if (configKey === 'resultReview' && target === 'signed') {
    const method = item.methodCode ? chain.methods[item.methodCode] : undefined;
    if (!method) blockers.push(`检测方法 ${item.methodCode || '未关联'} 不存在`);
    else if (!methodUsable(method.status)) blockers.push(`方法 ${method.code} 已失效（${method.status}），禁止签发`);
    const sample = item.sampleCode ? chain.samples[item.sampleCode] : undefined;
    if (!sample) blockers.push(`关联样本 ${item.sampleCode || '未关联'} 不存在`);
    else if (sample.status !== 'testing') blockers.push(`样本 ${sample.code} 状态为 ${sample.status}，须处于检测状态`);
    if (item.reviewRequestedBy && item.reviewRequestedBy === actor) blockers.push('复核人与签发人必须为不同人员');
  }
  return blockers;
}

File: backend/internal/service/chain.go
package service

import (
	"context"
	"fmt"
	"strings"

	"github.com/blueship581/water-sample-chain-assurance/backend/internal/constants"
	"github.com/blueship581/water-sample-chain-assurance/backend/internal/dto"
	"github.com/blueship581/water-sample-chain-assurance/backend/internal/repository"
)

// ChainService builds the release-chain view for the workbench: linked
// entities, live constraint checks and persisted block records, so a page
// refresh can always re-read the same chain state and blocking reasons.
type ChainService interface {
	Inspect(context.Context, string, uint) (dto.ChainView, error)
}

type chainService struct {
	batches  repository.SamplingBatchRepository
	samples  repository.LabSampleRepository
	methods  repository.AssayMethodRepository
	reviews  repository.ResultReviewRepository
	security SecurityService
}

func NewChainService(batches repository.SamplingBatchRepository, samples repository.LabSampleRepository, methods repository.AssayMethodRepository, reviews repository.ResultReviewRepository, security SecurityService) ChainService {
	return &chainService{batches: batches, samples: samples, methods: methods, reviews: reviews, security: security}
}

func (s *chainService) Inspect(ctx context.Context, entityType string, id uint) (dto.ChainView, error) {
	switch strings.TrimSpace(entityType) {
	case "sampling-batches":
		return s.inspectBatch(ctx, id)
	case "samples":
		return s.inspectSample(ctx, id)
	case "reviews":
		return s.inspectReview(ctx, id)
	default:
		return dto.ChainView{}, fmt.Errorf("%w: unsupported chain entity type %q", ErrInvalidInput, entityType)
	}
}

func (s *chainService) inspectBatch(ctx context.Context, id uint) (dto.ChainView, error) {
	batch, err := s.batches.Get(ctx, id)
	if err != nil {
		return dto.ChainView{}, err
	}
	samples, err := s.samples.ListByBatch(ctx, batch.Code, 20)
	if err != nil {
		return dto.ChainView{}, fmt.Errorf("list batch samples: %w", err)
	}
	links := make([]dto.ChainLink, 0, len(samples))
	undisposed := 0
	for _, sample := range samples {
		links = append(links, dto.ChainLink{
			Kind: "sample", Code: sample.Code, Name: sample.Name, Status: sample.Status, Exists: true,
		})
		if sample.Status != string(constants.SampleStateDisposed) {
			undisposed++
		}
	}
	closeCheck := dto.ChainCheck{Key: "no_undisposed_samples", Label: "关闭前样本均已处置", Passed: undisposed == 0}
	if undisposed > 0 {
		closeCheck.Reason = fmt.Sprintf("批次仍有 %d 份未处置样本，不得关闭", undisposed)
	}
	view := dto.ChainView{
		EntityType: "SamplingBatch", EntityID: batch.ID, Code: batch.Code, Status: batch.Status,
		Links: links, Checks: []dto.ChainCheck{closeCheck},
	}
	return s.attachBlocks(ctx, view, "SamplingBatch")
}

func (s *chainService) inspectSample(ctx context.Context, id uint) (dto.ChainView, error) {
	sample, err := s.samples.Get(ctx, id)
	if err != nil {
		return dto.ChainView{}, err
	}
	batchLink := dto.ChainLink{Kind: "batch", Code: sample.BatchCode}
	batchCheck := dto.ChainCheck{Key: "batch_received", Label: "所属批次已收到"}
	if batch, err := s.batches.GetByCode(ctx, sample.BatchCode); err != nil {
		batchLink.Exists = false
		batchCheck.Reason = fmt.Sprintf("所属批次 %s 不存在", sample.BatchCode)
	} else {
		batchLink.Exists, batchLink.Name, batchLink.Status = true, batch.Name, batch.Status
		batchCheck.Passed = batch.Status == string(constants.BatchStateReceived)
		if !batchCheck.Passed {
			batchCheck.Reason = fmt.Sprintf("批次 %s 当前状态为 %s，必须已收到(received)", batch.Code, batch.Status)
		}
	}
	methodLink := dto.ChainLink{Kind: "method", Code: sample.MethodCode}
	methodCheck := dto.ChainCheck{Key: "method_active", Label: "检测方法版本有效"}
	if method, err := s.methods.GetByCode(ctx, sample.MethodCode); err != nil {
		methodLink.Exists = false
		methodCheck.Reason = fmt.Sprintf("检测方法 %s 不存在", sample.MethodCode)
	} else {
		methodLink.Exists, methodLink.Name, methodLink.Status = true, method.Name, method.Status
		methodCheck.Passed = method.Status == string(constants.AssayMethodActive)
		if !methodCheck.Passed {
			methodCheck.Reason = fmt.Sprintf("方法 %s 当前状态为 %s，必须处于有效(active)版本", method.Code, method.Status)
		}
	}
	view := dto.ChainView{
		EntityType: "LabSample", EntityID: sample.ID, Code: sample.Code, Status: sample.Status,
		Links:  []dto.ChainLink{batchLink, methodLink},
		Checks: []dto.ChainCheck{batchCheck, methodCheck},
	}
	if sample.Status == string(constants.SampleStateDisposed) {
		view.Disposal = &dto.DisposalSnapshot{
			Reason: sample.DisposalReason, BatchCode: sample.DisposedBatchCode,
			MethodCode: sample.DisposedMethodCode, DisposedAt: sample.DisposedAt,
		}
	}
	return s.attachBlocks(ctx, view, "LabSample")
}

func (s *chainService) inspectReview(ctx context.Context, id uint) (dto.ChainView, error) {
	review, err := s.reviews.Get(ctx, id)
	if err != nil {
		return dto.ChainView{}, err
	}
	sampleLink := dto.ChainLink{Kind: "sample", Code: review.SampleCode}
	sampleCheck := dto.ChainCheck{Key: "sample_testing", Label: "样本处于检测状态"}
	if sample, err := s.samples.GetByCode(ctx, review.SampleCode); err != nil {
		sampleLink.Exists = false
		sampleCheck.Reason = fmt.Sprintf("关联样本 %s 不存在", review.SampleCode)
	} else {
		sampleLink.Exists, sampleLink.Name, sampleLink.Status = true, sample.Name, sample.Status
		sampleCheck.Passed = sample.Status == string(constants.SampleStateTesting)
		if !sampleCheck.Passed {
			sampleCheck.Reason = fmt.Sprintf("样本 %s 当前状态为 %s，必须处于检测状态(testing)", sample.Code, sample.Status)
		}
	}
	methodLink := dto.ChainLink{Kind: "method", Code: review.MethodCode}
	methodCheck := dto.ChainCheck{Key: "method_active", Label: "签发时方法仍有效"}
	if method, err := s.methods.GetByCode(ctx, review.MethodCode); err != nil {
		methodLink.Exists = false
		methodCheck.Reason = fmt.Sprintf("检测方法 %s 不存在", review.MethodCode)
	} else {
		methodLink.Exists, methodLink.Name, methodLink.Status = true, method.Name, method.Status
		methodCheck.Passed = method.Status == string(constants.AssayMethodActive)
		if !methodCheck.Passed {
			methodCheck.Reason = fmt.Sprintf("方法 %s 当前状态为 %s，必须处于有效(active)版本", method.Code, method.Status)
		}
	}
	reviewerCheck := dto.ChainCheck{
		Key: "reviewer_assigned", Label: "复核人与签发人分离",
		Passed: strings.TrimSpace(review.ReviewRequestedBy) != "",
	}
	if !reviewerCheck.Passed {
		reviewerCheck.Reason = "尚未提交复核，无法核对复核人与签发人分离"
	}
	view := dto.ChainView{
		EntityType: "ResultReview", EntityID: review.ID, Code: review.Code, Status: review.Status,
		Links:  []dto.ChainLink{sampleLink, methodLink},
		Checks: []dto.ChainCheck{methodCheck, sampleCheck, reviewerCheck},
	}
	return s.attachBlocks(ctx, view, "ResultReview")
}

// attachBlocks loads persisted release_blocked audit entries so blocked
// attempts remain readable after a page refresh.
func (s *chainService) attachBlocks(ctx context.Context, view dto.ChainView, entityType string) (dto.ChainView, error) {
	history, err := s.security.EntityHistory(ctx, entityType, view.EntityID, 50)
	if err != nil {
		return dto.ChainView{}, fmt.Errorf("load release blocks: %w", err)
	}
	blocks := make([]dto.ChainBlock, 0)
	for _, entry := range history {
		if entry.Action != "release_blocked" {
			continue
		}
		blocks = append(blocks, dto.ChainBlock{
			At: entry.CreatedAt, Actor: entry.Actor, Action: entry.Action, Reason: entry.Detail,
		})
	}
	view.Blocks = blocks
	return view, nil
}
